keps/sig-network/0034-20190125-ingress-api-group.md

```diff
@@ -0,0 +1,102 @@
+
+### Goals
+
+*   Move Ingress to a permanent API group.
+*   Position the Ingress API in a place towards making progress towards GA.
+
+## Proposal
+
+### 1.14
+
+*   Copy the Ingress API to networking.k8s.io/v1beta1 (preserving existing data and round-tripping with the extensions Ingress API, following the approach taken for all other extensions/v1beta1 resources)
+*   Develop a set of planned changes and GA graduation criteria with sig-network (intent is to target a minimal set of bugfixes and non-breaking changes)
+*   Announce extensions/v1beta1 Ingress as deprecated (and announce plan for GA)
+
+### 1.15
+
+*   Update API server to persist in networking.k8s.io/v1beta1
+*   Update in-tree controllers, examples, and clients to target networking.k8s.io/v1beta1
+*   Update Ingress controllers in the kubernetes org to target networking.k8s.io/v1beta1
+*   Update documentation to reference networking.k8s.io/v1beta1
+*   Work to meet graduation criteria and promote API to networking.k8s.io/v1
+*   Announce networking.k8s.io/v1beta1 Ingress as deprecated
+
+### 1.16
+
+*   Update API server to persist in networking.k8s.io/v1
+*   Update in-tree controllers, examples, and clients to target networking.k8s.io/v1
+*   Update Ingress controllers in the kubernetes org to target networking.k8s.io/v1
+*   Update documentation to reference networking.k8s.io/v1
+*   Evangelize availability of v1 Ingress API to out-of-org Ingress controllers
+
+### 1.18
+
+*   Remove ability to serve extensions/v1beta1 and networking.k8s.io/v1beta1 Ingress resources (preserve ability to read existing extensions/v1beta1 Ingress objects from storage and serve them via the networking.k8s.io/v1 API)
+
+## Graduation Criteria
+
+`networking.k8s.io/v1beta1`
+
+*   1.14: Ingress API exists and has parity with existing extensions/v1beta1 API
+*   1.14: extensions/v1beta1 Ingress tests are replicated against networking.k8s.io
+*   1.15: all in-tree use and in-org controllers switch to networking.k8s.io API group
+*   1.15: documentation and examples are updated to refer to networking.k8s.io API group
+
+`networking.k8s.io/v1`
+
+* TBD based on plans developed by sig-network in 1.14 timeframe
+
+## Implementation History
+
+TBD
+
+## Alternatives
+
+See motivation section.
+
+## Appendix
+
+### Non-options
+
+One suggestion was to move the API into a new API group, defined as a CRD.   This does not work because there is no way to do round-trip of existing Ingress objects to a CRD-based API.
```

Comment:
> round-trip of existing Ingress objects to a CRD-based API.

Is this a hard requirement? Why providing a way to dump Ingress object and convert them to the new type is not an option?

Reply:
I don't think that pushing this back on users is fair, especially if we're not substantially improving the API.  It seems very likely to break for someone (e.g. they don't get the memo), and we have a LOT of someones using k8s now :)

Reply:
Agreed. I think the adoption is too high for outright deletion (which is what a move to a CRD would involve). Providing a path forward and 9-12 months of overlap between old and new API groups (1.14-1.18 is proposed) is a much gentler transition.